src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,328 @@
+            else
+            {
+                CoreEventSource.Log.FailedToSetHeartbeatProperty(healthPropertyName, healthPropertyValue, isHealthy.HasValue ? isHealthy.Value.ToString() : "null");
+            }
+
+            return false;
+        }
+
+        #region IDisposable Support
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        /// <summary>
+        /// Get the metric telemetry item that will be sent.
+        /// 
+        /// Note: exposed to internal to allow inspection for testing.
+        /// </summary>
+        /// <returns>A MetricTelemtry item that contains the currently defined payload for a heartbeat 'pulse'.</returns>
+        internal ITelemetry GatherData()
+        {
+            var hbeat = new MetricTelemetry(heartbeatSyntheticMetricName, 0.0);
+
+            string updatedKeys = string.Empty;
+            string comma = string.Empty;
+
+            if (hbeat.Properties.ContainsKey(HeartbeatDefaultPayload.UpdatedFieldsPropertyKey))
+            {
+                updatedKeys = hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey];
+                comma = ",";
+            }
+
+            foreach (var payloadItem in this.heartbeatProperties)
+            {
+                hbeat.Properties.Add(payloadItem.Key, payloadItem.Value.PayloadValue);
+                hbeat.Sum += payloadItem.Value.IsHealthy ? 0 : 1;
+                if (payloadItem.Value.IsUpdated)
+                {
+                    string.Concat(updatedKeys, comma, payloadItem.Key);
+                    comma = ",";
+                }
+            }
+
+            // update the special 'updated keys' property with the names of keys that have been updated.
+            if (!string.IsNullOrEmpty(updatedKeys))
+            {
+                hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey] = updatedKeys;
+            }
+
+            hbeat.Sequence = string.Format(CultureInfo.CurrentCulture, "{0}", this.heartbeatsSent++);
+
+            return hbeat;
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                    if (this.HeartbeatTimer != null)
+                    {
+                        this.isEnabled = false;
+
+                        try
+                        {
+                            this.HeartbeatTimer.Dispose();
+                        }
+                        catch (Exception)
+                        {
+                            CoreEventSource.Log.LogError("Error occured when disposing heartbeat timer within HeartbeatProvider");
+                        }
+                    }
+                }
+
+                this.disposedValue = true;
+            }
+        }
+
+        #endregion
+
+        private void Send()
+        {
+            if (this.telemetryClient.TelemetryConfiguration.TelemetryChannel == null)
```

Comment:
how is this possible?

Reply:
[The same way it would be possible in PortalDiagnosticsSender](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/7321d2bf7b9c4733fb43b6c186e381863931eb1a/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/PortalDiagnosticsSender.cs#L93). This is where I learned how we should send things along, and if I am not mistaken it is not possible to create a TelemetryConfiguration with a null TelemetryChannel, but it is possible to set that TelemetryChannel to null afterward. I looked through the calls that would be sent outside of my use case and I see that it is indeed possible to end up in this scenario, but not entirely likely.